tls/cfsslscan_handshake.go

```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
 	if err != nil {
 		return
 	}
+	if CipherSuites[serverHello.cipherSuite].EllipticCurve {
+		// Prime the connection, if necessary, for key
+		// exchange messages by reading off the certificate
+		// message and, if necessary, the OCSP stapling
+		// message
+		var msg interface{}
+		msg, err = c.readHandshake()
+		if err != nil {
+			return
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
+			msg, err = c.readHandshake()
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
+			if !ok {
+				c.sendAlert(alertUnexpectedMessage)
+				return
+			}
+		}
+
+		var skx *serverKeyExchangeMsg
+		skx, err = c.exchangeKeys()
+		if skx.raw[0] != typeServerKeyExchange {
+			err = errors.New("tls: received something other than the expected ServerKeyExchange.")
+			return
+		}
+		if err != nil {
```

Comment:
This is pretty standard in Go, minimal checking followed by short circuiting that passes errors up as high on the stack as possible before being handled with some action (logging, sending messages, stopping the goroutine, etc.)